### PR TITLE
Refactor Meter to Map code in ValueRefs.cpp

### LINKED ISSUE #4060: Refactor Meter to Map code in ValueRefs.cpp
Follow-up to a side discussion started in: https://github.com/freeorion/freeorion/issues/4047

Originally suggested by @geoffthemedio 
> -refactor GetMeterNameMap, NameToMeter, and MeterToName in ValueRef.cpp to use a constexpr array of pair, or two constexpr arrays to hold the strings (as string_view) and the MeterTypes, or a single constexpr array indexed by MeterTypes converted to suitable values (accounting for INVALID_METER_TYPE = -1), rather than the current use of a static map<string, MeterType>

I'm creating an issue ticket for this so it's clear someone's working on it. if it's preferable to simply open a draft PR instead, I will close this ticket and open a PR.

### PR DESCRIPTION
**Pull request by AndyMender**
fixes #4060

### CODE REVIEW

File: universe/ValueRefs.cpp [hunk 1]
```diff
@@ -234,50 +235,45 @@ namespace {
...
-            {"SecondaryStat",        MeterType::METER_SECONDARY_STAT},
-            {"MaxSecondaryStat",     MeterType::METER_MAX_SECONDARY_STAT},
-            {"Size",                 MeterType::METER_SIZE}
-        };
-        return meter_name_map;
-    }
-    // force early init to avoid threading issues later
-    std::map<std::string, MeterType> dummy = GetMeterNameMap();
-
+    const std::array<std::pair<std::string, MeterType>, 36> NAME_TO_METER = {{
```
Comment (AndyMender) [inline universe/ValueRefs.cpp]: I tried using a `constexpr` array with `std::string_view`, however the issue is `MeterToName` and the boost::phoenix wrapper which fails during linking if the return type is `std::string_view`.

File: universe/ValueRefs.cpp [hunk 2]
```diff
@@ -310,20 +306,24 @@ std::string ValueRefBase::InvariancePattern() const {
...
+    auto it = std::find_if(NAME_TO_METER.begin(), NAME_TO_METER.end(),
+        [&meter](const std::pair<std::string, MeterType>& elem) {
+            return elem.second == meter;
+        }
+    );
+
+    return (it != NAME_TO_METER.end()) ? it->first : EMPTY_STRING;
 }
 
 constexpr std::string_view PlanetTypeToStringConstexpr(PlanetType type) {
```
Comment (AndyMender) [inline universe/ValueRefs.cpp:307]: I can later have a look if it's possible to simplify the PlanetType code here to also use an array of pairs.

File: universe/ValueRefs.cpp (hunk 1 above)
Comment (o01eg) [reply universe/ValueRefs.cpp]: What is boost::phoenix issue? Maybe it could be solved with new Python FOCS parser?

File: universe/UniverseObject.cpp [hunk 3]
```diff
@@ -16,7 +16,7 @@
 #include "../util/i18n.h"
 
 namespace ValueRef {
-    const std::string& MeterToName(MeterType meter);
+    const std::string& MeterToName(const MeterType& meter);
```
Comment (geoffthemedio) [inline universe/UniverseObject.cpp]: `MeterType` is an enum so is better passed by value in cases like this

File: universe/ValueRefs.cpp (hunk 1 above)
Comment (geoffthemedio) [reply universe/ValueRefs.cpp]: probably https://github.com/freeorion/freeorion/blob/master/parse/DoubleComplexValueRefParser.cpp#L10

I think a locally defined wrapper would resolve that...

File: universe/ValueRefs.cpp [hunk 4]
```diff
@@ -308,22 +308,24 @@ std::string ValueRefBase::InvariancePattern() const {
         .append(ConstantExpr()                  ? "C" : "c");
 }
 
-MeterType NameToMeter(const std::string& name) {
-    auto it = std::find_if(METER_TO_NAME.begin(), METER_TO_NAME.end(),
+MeterType NameToMeter(const std::string_view name) {
```
Comment (geoffthemedio) [inline universe/ValueRefs.cpp]: can pass `string_view` by value

File: universe/ValueRefs.cpp [hunk 5]
```diff
@@ -308,22 +308,24 @@ std::string ValueRefBase::InvariancePattern() const {
...
+    return (it != NAME_TO_METER.end()) ? it->second : MeterType::INVALID_METER_TYPE;
 }
 
 const std::string& MeterToName(const MeterType& meter) {
-    for (auto& [name, type] : METER_TO_NAME) {
-        if (type == meter)
-            return std::string{name}; // CodeQL reports "Returning stack-allocated memory", but the name structured binding should be a reference as far as I can tell
-    }
-    return EMPTY_STRING;
+    auto it = std::find_if(NAME_TO_METER.begin(), NAME_TO_METER.end(),
```
Comment (geoffthemedio) [inline universe/ValueRefs.cpp]: Part of the suggestion was to replace this linear search with `find_if` with an index into an array. So instead of storing an `array<pair<string_view, MeterType>>`, you can convert the `MeterType` to an index (accounting for `INVALID_METER_TYPE = -1` by offsetting the value before casting to `std::size_t`) and just use the index to get the name from the `array<string_view>` as `array[index]`.

File: universe/ValueRefs.cpp (hunk 4 above)
Comment (AndyMender) [reply universe/ValueRefs.cpp]: Isn't it already passed by value?

File: universe/UniverseObject.cpp (hunk 3 above)
Comment (AndyMender) [reply universe/UniverseObject.cpp]: Good point, changed!

**Comment by hifron** [comment]
Needs `git rebase` to update to current git master(and there is [GitHub Desktop GUI](https://desktop.github.com/) for that for those who don't want to remember commands which are not so simple(and FreeOrion project policy to not import FO master branch into pull branch(therefore master need `git pull` and then switch branch and do `git rebase` on master(easy with GUI(but needs on your web GitHub master branch https://github.com/AndyMender/freeorion to click to sync FO project to not be behind)))))...

If pull branch is correct after rebase could be seen in GUI history or with `git log` and after git pushing could also be seen on web pull branch fix-4060 (in header of this page is link)(but _on web of this pull request there should be no web update branch click_, because this also import FO master branch into pull branch(and FO this don't want))

File: universe/ValueRefs.cpp (hunk 5 above)
Comment (AndyMender) [reply universe/ValueRefs.cpp]: Instead of offsetting I did the following:
```cpp
const std::string& MeterToName(const MeterType meter) {
    return (meter == MeterType::INVALID_METER_TYPE) ? EMPTY_STRING : NAME_BY_METER[static_cast<std::size_t>(meter)];
}
```

File: universe/ValueRefs.cpp (hunk 4 above)
Comment (geoffthemedio) [reply universe/ValueRefs.cpp]: yes

File: universe/ValueRefs.cpp (hunk 5 above)
Comment (geoffthemedio) [reply universe/ValueRefs.cpp]: probably OK in practice, though potentially introduces an unnecessary branch that the offsetting variant could avoid.

Comment (AndyMender) [reply universe/ValueRefs.cpp]: Went with offsetting in the end.

File: universe/ValueRefs.cpp [hunk 6]
```diff
@@ -235,8 +235,11 @@ namespace {
         mutable UniverseObjectType m_type;
     };
 
-    // Array of meter names enumerated by MeterType
-    const std::array<std::string, static_cast<std::size_t>(MeterType::NUM_METER_TYPES)> NAME_BY_METER = {
+    const std::string EMPTY_STRING;
+
+    // Array of meter names enumerated by MeterType with INVALID_METER_TYPE as first element
+    const std::array<std::string, static_cast<std::size_t>(MeterType::NUM_METER_TYPES) + 1> NAME_BY_METER = {
+        EMPTY_STRING,
```
Comment (geoffthemedio) [inline universe/ValueRefs.cpp]: don't need `EMPTY_STRING` here... can just write `""`

File: universe/ValueRefs.cpp [hunk 7]
```diff
@@ -316,14 +318,14 @@ MeterType NameToMeter(const std::string_view name) {
     for (std::size_t i = 0; i < NAME_BY_METER.size(); i++)
     {
         if (NAME_BY_METER[i] == name)
-            return static_cast<MeterType>(i);
+            return static_cast<MeterType>(i - 1);
```
Comment (geoffthemedio) [inline universe/ValueRefs.cpp:292]: are you sure this works? `i` is `size_t` which will probably underflow to a big number, rather than the -1 you probably want if `name` is an empty string so that `i = 0`. not sure what happens if you cast that to `MeterType` with underlying type `signed char`.

File: universe/ValueRefs.cpp [hunk 8]
```diff
@@ -316,14 +318,14 @@ MeterType NameToMeter(const std::string_view name) {
...
     }
 
     return MeterType::INVALID_METER_TYPE;
 }
 
 const std::string& MeterToName(const MeterType meter) {
-    return (meter == MeterType::INVALID_METER_TYPE) ? EMPTY_STRING : NAME_BY_METER[static_cast<std::size_t>(meter)];
+    return NAME_BY_METER[static_cast<int>(meter) + 1];
```
Comment (geoffthemedio) [inline universe/ValueRefs.cpp]: probably better cast to `std::underlying_type_t<MeterType>`

File: universe/ValueRefs.cpp (hunk 7 above)
Comment (geoffthemedio) [reply universe/ValueRefs.cpp:292]: ideally `NAME_BY_METER` would be `constexpr` so you could `static_assert` to make sure the results are as expected

File: universe/ValueRefs.cpp (hunk 1 above)
Comment (AndyMender) [reply universe/ValueRefs.cpp]:
```cpp
    std::string MeterToNameWrapper(MeterType meter) {
        return std::string().append(ValueRef::MeterToName(meter));
    }

    BOOST_PHOENIX_ADAPT_FUNCTION(std::string, MeterToName_, MeterToNameWrapper, 1)
```

I defined the above wrapper. I'm not familiar with `BOOST_PHOENIX_ADAPT_FUNCTION`, but would a lambda work as well or does it have to be a named function?

File: universe/ValueRefs.cpp (hunk 6 above)
Comment (AndyMender) [reply universe/ValueRefs.cpp]: Removed 👍🏼

File: universe/ValueRefs.cpp (hunk 7 above)
Comment (AndyMender) [reply universe/ValueRefs.cpp:292]: Whether the original operation works would probably be implementation depend so no need to risk it. I switched to regular `int`.

Should the `static_assert` check if `i - 1` is within allowed enum range?

File: universe/ValueRefs.cpp (hunk 8 above)
Comment (AndyMender) [reply universe/ValueRefs.cpp]: Yup, changed.

File: universe/ValueRefs.cpp (hunk 7 above)
Comment (hifron) [reply universe/ValueRefs.cpp:292]: Tests are failing, _probably_ because array starts with position 0 or some reference pointer(string_view?) mismatch.

Comment (AndyMender) [reply universe/ValueRefs.cpp:292]: Looks like it's failing on `METER_POPULATION` which has index 16. Not sure why and unfortunately I don't have a Windows machine to debug locally. Checking other code paths which may result in a `0.0` return in the test.

Comment (AndyMender) [reply universe/ValueRefs.cpp:292]: Fixed that part.

File: UI/ObjectListWnd.cpp [hunk 9]
```diff
@@ -310,7 +310,7 @@ namespace {
             for (MeterType meter = MeterType(0); meter <= MeterType::METER_SPEED;  // the meter(s) after MeterType::METER_SPEED are part-specific
                  meter = MeterType(int(meter) + 1))
             {
-                col_types[{std::string{to_string(meter)},           UserStringNop("METERS_SUBMENU")}] = StringCastedImmediateValueRef(ValueRef::MeterToName(meter));
+                col_types[{std::string{to_string(meter)},           UserStringNop("METERS_SUBMENU")}] = StringCastedImmediateValueRef(std::string().append(ValueRef::MeterToName(meter)));
```
Comment (geoffthemedio) [inline UI/ObjectListWnd.cpp]:
```suggestion
                col_types[{std::string{to_string(meter)},           UserStringNop("METERS_SUBMENU")}] = StringCastedImmediateValueRef(std::string{ValueRef::MeterToName(meter)});
```

File: parse/DoubleComplexValueRefParser.cpp [hunk 10]
```diff
@@ -7,7 +7,11 @@
 
 namespace parse {
 
-    BOOST_PHOENIX_ADAPT_FUNCTION(std::string, MeterToName_, ValueRef::MeterToName, 1)
+    std::string MeterToNameWrapper(MeterType meter) {
+        return std::string().append(ValueRef::MeterToName(meter));
+    }
```
Comment (geoffthemedio) [inline parse/DoubleComplexValueRefParser.cpp]:
```suggestion
    std::string MeterToNameWrapper(MeterType meter)
    { return std::string{ValueRef::MeterToName(meter)}; }
```

File: universe/UniverseObject.cpp [hunk 11]
```diff
@@ -16,7 +16,7 @@
 #include "../util/i18n.h"
 
 namespace ValueRef {
-    const std::string& MeterToName(MeterType meter);
+    const std::string_view MeterToName(const MeterType meter);
```
Comment (geoffthemedio) [inline universe/UniverseObject.cpp]:
```suggestion
    std::string_view MeterToName(const MeterType meter);
```

File: universe/ValueRefs.cpp [hunk 12]
```diff
@@ -285,20 +286,18 @@ std::string ValueRefBase::InvariancePattern() const {
...
 
-MeterType NameToMeter(const std::string& name) {
-    MeterType retval = MeterType::INVALID_METER_TYPE;
-    auto it = GetMeterNameMap().find(name);
-    if (it != GetMeterNameMap().end())
-        retval = it->second;
-    return retval;
+MeterType NameToMeter(const std::string_view name) {
+    for (int i = 0; i < static_cast<int>(NAME_BY_METER.size()); i++)
+    {
```
Comment (geoffthemedio) [inline universe/ValueRefs.cpp]:
```suggestion
    for (int i = 0; i < static_cast<int>(NAME_BY_METER.size()); i++) {
```

File: universe/ValueRefs.cpp (hunk 7 above)
Comment (geoffthemedio) [reply universe/ValueRefs.cpp:292]: > Should the `static_assert` check if `i - 1` is within allowed enum range?''

Rather that `NameToMeter("Population") == MeterType::METER_POPULATION`, `NameToMeter("not a meter") == MeterType::INVALID_METER_TYPE`, and similar. Would need to make NameToMeter `constexpr`, which hopefully wouldn't be complicated by it being separately declared in a header. Also that `MeterToName` calls produce expected results at the extremes.

File: universe/ValueRefs.cpp [hunk 13]
```diff
@@ -285,20 +286,18 @@ std::string ValueRefBase::InvariancePattern() const {
...
 }
 
-const std::string& MeterToName(MeterType meter) {
-    for (auto& [name, type] : GetMeterNameMap()) {
-        if (type == meter)
-            return name; // CodeQL reports "Returning stack-allocated memory", but the name structured binding should be a reference as far as I can tell
-    }
-    return EMPTY_STRING;
+const std::string_view MeterToName(const MeterType meter) {
+    return NAME_BY_METER[static_cast<std::underlying_type_t<MeterType>>(meter) + 1];
```
Comment (geoffthemedio) [inline universe/ValueRefs.cpp:304]: I'd add a comment here noting that the lowest value for `meter` should be `INVALID_METER_TYPE = -1` and the highest less than `NUM_METER_TYPES`

File: universe/ValueRefs.h [hunk 14]
```diff
@@ -448,8 +448,8 @@ struct FO_COMMON_API Operation final : public ValueRef<T>
 
 /* Convert between names and MeterType. Names are scripting token, like Population
  * and not the MeterType string representations like METER_POPULATION */
-[[nodiscard]] FO_COMMON_API MeterType          NameToMeter(const std::string& name);
-[[nodiscard]] FO_COMMON_API const std::string& MeterToName(MeterType meter);
+[[nodiscard]] FO_COMMON_API MeterType          NameToMeter(const std::string_view name);
+[[nodiscard]] FO_COMMON_API const std::string_view MeterToName(const MeterType meter);
```
Comment (geoffthemedio) [inline universe/ValueRefs.h]:
```suggestion
[[nodiscard]] FO_COMMON_API std::string_view MeterToName(const MeterType meter);
```
`const` of by-value return type is unnecessary. maybe also could impair return value optimization, but I'm just guessing.

**Comment by hifron** [comment]
(Not sure if required for tests)
`pip install -r default/python/requirements-dev.txt`
Under Windows(compiled with cmake, required to -DBUILD_TESTING=On during VS solution generation) to run tests:
```
cd build
cmake --build . --config "Release" --target unittest
```

[VS Code(Not Full VisualStudio)](https://code.visualstudio.com/) regenerates cmake configuration on each change or on startup, therefore option(BUILD_TESTING "Build the testing tree." ON) in [CMakeLists.txt](https://github.com/freeorion/freeorion/blob/master/CMakeLists.txt) is better

Other aspects of testing are under [.github directory](https://github.com/freeorion/freeorion/tree/master/.github), not all platforms have tests enabled by default in GitHub FO CI checks.

File: universe/ValueRefs.h (hunk 14 above)
Comment (AndyMender) [reply universe/ValueRefs.h]: I think this might've been the source of the initial compilation issues I ran into. This allowed me to use the `std::string` constructor with `string_view`values.

File: universe/ValueRefs.cpp (hunk 13 above)
Comment (AndyMender) [reply universe/ValueRefs.cpp:304]: Done 👍🏼

File: universe/ValueRefs.cpp (hunk 7 above)
Comment (AndyMender) [reply universe/ValueRefs.cpp:292]: Added static asserts for `NameToMeter`. Still need to figure out the issue with constness in `NameToMeter` and why it can't easily be `constexpr` as well.

File: universe/ValueRefs.cpp [hunk 15]
```diff
@@ -853,7 +856,7 @@ double Variable<double>::Eval(const ScriptingContext& context) const
 {
     IF_CURRENT_VALUE(double)
 
-    const std::string& property_name = m_property_name.empty() ? "" : m_property_name.back();
+    const std::string property_name = m_property_name.empty() ? "" : m_property_name.back();
```
Comment (geoffthemedio) [inline universe/ValueRefs.cpp]: why this change?

Comment (AndyMender) [reply universe/ValueRefs.cpp]: Oversight, reverted.

File: universe/ValueRefs.cpp (hunk 7 above)
Comment (AndyMender) [reply universe/ValueRefs.cpp:292]: For `MeterToName` I'm running into the following errors:
```cpp
[225/285] Building CXX object CMakeFiles/freeorion.dir/UI/ObjectListWnd.cpp.o
In file included from /home/amender/projects/freeorion/UI/ObjectListWnd.cpp:25:
/home/amender/projects/freeorion/UI/../universe/ValueRefs.h:452:59: warning: inline function ‘constexpr std::string_view ValueRef::MeterToName(MeterType)’ used but never defined
  452 | [[nodiscard]] FO_COMMON_API constexpr std::string_view    MeterToName(const MeterType meter);
```
It seems like either `MeterType` is not treated as a literal type or I can't force values passed to `MeterToName` to be const-correct.

Comment (hifron) [reply universe/ValueRefs.cpp:292]: Isn't constexpr MeterType, not string_view ? (because in this case we want to evaluate MeterType and return something to GUI?)

File: universe/ValueRefs.cpp [hunk 16]
```diff
@@ -285,20 +286,22 @@ std::string ValueRefBase::InvariancePattern() const {
...
 
-const std::string& MeterToName(MeterType meter) {
-    for (auto& [name, type] : GetMeterNameMap()) {
-        if (type == meter)
-            return name; // CodeQL reports "Returning stack-allocated memory", but the name structured binding should be a reference as far as I can tell
-    }
-    return EMPTY_STRING;
+static_assert(NameToMeter("not a meter") == MeterType::INVALID_METER_TYPE, "Name to Meter conversion failed for invalid meter type!");
+static_assert(NameToMeter("Population") == MeterType::METER_POPULATION, "Name to Meter conversion failed for 'Population' meter!");
+static_assert(NameToMeter("Speed") == MeterType::METER_SPEED, "Name to Meter conversion failed for 'Speed' meter!");
```
Comment (hifron) [inline universe/ValueRefs.cpp:300]: Seem strange that such things are needed that without it it fails.

**Review by hifron — COMMENTED** [review]
Seems strange that such things are needed.

File: universe/ValueRefs.cpp (hunk 16 above)
Comment (AndyMender) [reply universe/ValueRefs.cpp:300]: This is mostly future proofing if more meters are added and/or the enum order changes I think.

Comment (hifron) [reply universe/ValueRefs.cpp:300]: Such things should be in tests and current implementation is error prone in future code edit.

Comment (geoffthemedio) [reply universe/ValueRefs.cpp:300]: There is nothing wrong with using `static_assert` for testing when appropriate. If a code edit breaks it, then the test will fail.

Comment (hifron) [reply universe/ValueRefs.cpp:300]: In [TestEnumParser.cpp](https://github.com/freeorion/freeorion/blob/master/test/parse/TestEnumParser.cpp) there are tests for each Enum like:
```
CHECK_ENUM_AND_RESULT("TargetPopulation", MeterType::METER_TARGET_POPULATION, MeterType, parse::non_ship_part_meter_enum_grammar);
CHECK_ENUM_AND_RESULT("SetTargetPopulation", MeterType::METER_TARGET_POPULATION, MeterType, parse::set_non_ship_part_meter_enum_grammar);
CHECK_ENUM_AND_RESULT("TargetIndustry", MeterType::METER_TARGET_INDUSTRY, MeterType, parse::non_ship_part_meter_enum_grammar);
CHECK_ENUM_AND_RESULT("SetTargetIndustry", MeterType::METER_TARGET_INDUSTRY, MeterType, parse::set_non_ship_part_meter_enum_grammar);
CHECK_ENUM_AND_RESULT("TargetResearch", MeterType::METER_TARGET_RESEARCH, MeterType, parse::non_ship_part_meter_enum_grammar);
CHECK_ENUM_AND_RESULT("SetTargetResearch", MeterType::METER_TARGET_RESEARCH, MeterType, parse::set_non_ship_part_meter_enum_grammar);
CHECK_ENUM_AND_RESULT("TargetInfluence", MeterType::METER_TARGET_INFLUENCE, MeterType, parse::non_ship_part_meter_enum_grammar);
CHECK_ENUM_AND_RESULT("SetTargetInfluence", MeterType::METER_TARGET_INFLUENCE, MeterType, parse::set_non_ship_part_meter_enum_grammar);
CHECK_ENUM_AND_RESULT("TargetConstruction", MeterType::METER_TARGET_CONSTRUCTION, MeterType, parse::non_ship_part_meter_enum_grammar);
CHECK_ENUM_AND_RESULT("SetTargetConstruction", MeterType::METER_TARGET_CONSTRUCTION, MeterType, parse::set_non_ship_part_meter_enum_grammar);
CHECK_ENUM_AND_RESULT("MaxFuel", MeterType::METER_MAX_FUEL, MeterType, parse::non_ship_part_meter_enum_grammar);
CHECK_ENUM_AND_RESULT("SetMaxFuel", MeterType::METER_MAX_FUEL, MeterType, parse::set_non_ship_part_meter_enum_grammar);
CHECK_ENUM_AND_RESULT("MaxShield", MeterType::METER_MAX_SHIELD, MeterType, parse::non_ship_part_meter_enum_grammar);
CHECK_ENUM_AND_RESULT("SetMaxShield", MeterType::METER_MAX_SHIELD, MeterType, parse::set_non_ship_part_meter_enum_grammar);
CHECK_ENUM_AND_RESULT("MaxStructure", MeterType::METER_MAX_STRUCTURE, MeterType, parse::non_ship_part_meter_enum_grammar);
CHECK_ENUM_AND_RESULT("SetMaxStructure", MeterType::METER_MAX_STRUCTURE, MeterType, parse::set_non_ship_part_meter_enum_grammar);
CHECK_ENUM_AND_RESULT("MaxDefense", MeterType::METER_MAX_DEFENSE, MeterType, parse::non_ship_part_meter_enum_grammar);

```

In [TestValueRefDoubleParser.cpp](https://github.com/freeorion/freeorion/blob/master/test/parse/TestValueRefDoubleParser.cpp) there is example of parsing math example and result of AbstractSyntaxTree which tell compiler how to know what it means:
```
/*
Term:
  -1.5+2.9/-7.4

Expected AST:
  PLUS
    NEGATE
      1.5
    DIVIDE
      2.9
      NEGATE
        7.4
*/
```

Such tests should also be doable for something **as seen in [TestEnumParser.cpp](https://github.com/freeorion/freeorion/blob/master/test/parse/TestEnumParser.cpp) comments**(and when that happens, position in array should be problem, but tests of enum are in place and missing should be added in future(**but adding it in wrong order cause test to fail as it should**))´:
```
// XXX: METER_SIZE not handled, still used?
// XXX: HAPPINESS meters not handled, still used?
// XXX: REBEL_TROOPS meters not handled, still used?
// XXX: No enum number value to validate enum coverage.
//      Maybe the Meter enum should be split.
```

Therefore AbstractSyntaxTree for something like([Python part tech](https://github.com/freeorion/freeorion/blob/master/default/scripting/techs/ship_parts/fuel/fuel.py)):
```
def PART_UPGRADE_MAXFUEL_EFFECTS(effect_name: str, value: float):
    return EffectsGroup(
        scope=Ship & OwnedBy(empire=Source.Owner) & DesignHasPart(name="FU_BASIC_TANK"),
        accountinglabel=effect_name,
        effects=[
            SetMaxFuel(
                value=Value
                + NamedReal(name=effect_name + "_MULT", value=value)
                * PartsInShipDesign(name="FU_BASIC_TANK", design=Target.DesignID)
            )
        ],
    )

```

SetMaxFuel:
 previous_value + some_effect * ship_parts_capacity


When wanting in [ValueRefs.cpp](https://github.com/freeorion/freeorion/blob/master/universe/ValueRefs.cpp) convert and evaluate some string_view to do some (meter) type and viceversa, there are also **invariants** (=something that is always true)(or candidate)

**For a function to be fit for use in constant expressions, it must be explicitly declared constexpr**. From [What's the difference between constexpr and const?](https://stackoverflow.com/questions/14116003/whats-the-difference-between-constexpr-and-const)
